File: examples/csharp/mega_flow/Pipeline.cs
using System;
using System.Collections.Generic;
using System.Linq;
using static System.String;
using Tasks = System.Threading.Tasks;

namespace Mega;

// Pipeline — exercises C#'s idiomatic flow constructs: LINQ, lambdas,
// delegates, async/await, try/catch/finally, switch expressions,
// pattern matching, records, tuples, yield iterators.
public static class Pipeline
{
    public static async Tasks.Task<string> OrchestrateAsync(Envelope envelope)
    {
        await Tasks.Task.Yield();
        return Orchestrate(envelope);
    }

    // Delegate factory — returns a Func reducer that joins tokens.
    private static Func<string, string, string> MakeJoiner(string sep)
        => (acc, tok) => string.IsNullOrEmpty(acc) ? tok : $"{acc}{sep}{tok}";

    // yield iterator — streams every tainted token.
    private static IEnumerable<string> Tokenize(string cmd)
    {
        foreach (var part in cmd.Split(' '))
        {
            if (!string.IsNullOrEmpty(part))
                yield return part;
        }
    }

    public static string Orchestrate(Envelope envelope)
    {
        var cmd = envelope.Cmd;
        var user = envelope.User;
        string? maybeUser = envelope.User;
        if (envelope is { Kind: Kind.Run })
        {
            user = maybeUser ?? user;
        }
        Tuple<string, int> legacyTuple = Tuple.Create(cmd, cmd.Length);
        ValueTuple<string, string> namedTuple = (legacyTuple.Item1, user);
        cmd = namedTuple.Item1;

        // LINQ pipeline: Select → Where → Aggregate with a closure.
        var joiner = MakeJoiner(" ");
        var joined = Tokenize(cmd)
            .Select(s => s.Trim())
            .Where(s => s.Length > 0)
            .Aggregate("", joiner);

        // Switch expression — every arm preserves taint.
        var routed = envelope.Kind switch
        {
            Kind.Run  => $"{joined}",
            Kind.Eval => joined.Trim(),
            _         => joined,
        };

        // try / catch / finally — taint survives every branch.
        Envelope valid;
        using (var audit = new AuditScope())
        {
            try
            {
                if (IsNullOrEmpty(routed)) throw new InvalidOperationException("empty");
                valid = envelope with { Cmd = routed, User = user, Length = routed.Length };
            }
            catch
            {
                valid = envelope with { Cmd = routed, User = user, Length = routed.Length };
            }
            finally
            {
                audit.Dispose();
            }
        }

        return Storage.Persist(valid);
    }

    private sealed class AuditScope : IDisposable
    {
        public void Dispose() { }
    }
}

File: examples/rust/mega_flow/src/pipeline.rs
use crate::storage as store;
use crate::{Envelope, Kind};

// Pipeline — exercises Rust's idiomatic flow constructs: iterators,
// closures, match expressions, Result / Option, `?`, generics, traits,
// if-let, while-let.

// Closure factory — returns a boxed Fn reducer that joins tokens.
fn make_joiner(sep: &'static str) -> Box<dyn Fn(String, &str) -> String> {
    Box::new(move |acc: String, tok: &str| -> String {
        if acc.is_empty() { tok.to_string() } else { format!("{}{}{}", acc, sep, tok) }
    })
}

pub fn orchestrate(envelope: Envelope) -> String {
    let cmd = envelope.cmd.clone();
    let user = envelope.user.clone();

    // Iterator chain — map / filter / fold with a closure.
    let joiner = make_joiner(" ");
    let joined: String = cmd
        .split_whitespace()
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .fold(String::new(), |acc, tok| joiner(acc, tok));

    // Match expression — every arm preserves taint.
    let routed: String = match envelope.kind {
        Kind::Run => format!("{}", joined),
        Kind::Eval => joined.trim().to_string(),
    };
    let routed = if let Some(value) = Some(routed) {
        value
    } else {
        String::new()
    };
    let mut pending = vec![routed.clone()];
    while let Some(value) = pending.pop() {
        if value.is_empty() {
            break;
        }
    }

    // Result style — taint survives the branch via closures.
    let valid: Envelope = (|| -> Result<Envelope, &str> {
        if routed.is_empty() { return Err("empty"); }
        Ok(Envelope {
            kind: envelope.kind.clone(),
            cmd: routed.clone(),
            user: user.clone(),
            length: routed.len(),
            extras: envelope.extras.clone(),
        })
    })().unwrap_or_else(|_| Envelope {
        kind: envelope.kind.clone(),
        cmd: routed.clone(),
        user: user.clone(),
        length: routed.len(),
        extras: envelope.extras.clone(),
    });

    store::persist(valid)
}
